pgmpy/inference/Sampling.py

```diff
@@ -0,0 +1,274 @@
+        """
+        Generates weighted sample(s) from joint distribution of the bayesian
+        network, that comply with the given evidence.
+        'Probabilistic Graphical Model Principles and Techniques', Koller and
+        Friedman, Algorithm 12.2 pp 493.
+
+        Parameters
+        ----------
+        evidence: dict
+            a dict key, value pair as {var: state_of_var_observed}
+            None if no evidence
+        size: int
+            size of sample to be generated
+
+        Returns
+        -------
+        sampled: pandas.DataFrame
+            the generated samples with corresponding weights
+
+        Examples
+        --------
+        >>> from pgmpy.factors.Factor import State
+        >>> from pgmpy.models.BayesianModel import BayesianModel
+        >>> from pgmpy.factors.CPD import TabularCPD
+        >>> from pgmpy.inference.Sampling import BayesianModelSampling
+        >>> student = BayesianModel([('diff', 'grade'), ('intel', 'grade')])
+        >>> cpd_d = TabularCPD('diff', 2, [[0.6], [0.4]])
+        >>> cpd_i = TabularCPD('intel', 2, [[0.7], [0.3]])
+        >>> cpd_g = TabularCPD('grade', 3, [[0.3, 0.05, 0.9, 0.5], [0.4, 0.25,
+        ...         0.08, 0.3], [0.3, 0.7, 0.02, 0.2]],
+        ...         ['intel', 'diff'], [2, 2])
+        >>> student.add_cpds(cpd_d, cpd_i, cpd_g)
+        >>> inference = BayesianModelSampling(student)
+        >>> evidence = {'diff': State(var='diff',state=0)}
+        >>> inference.likelihood_weighted_sample(evidence, 2)
+                intel       diff       grade  _weight
+        0  (intel, 0)  (diff, 0)  (grade, 1)      0.6
+        1  (intel, 1)  (diff, 0)  (grade, 1)      0.6
+        """
+        sampled = DataFrame(index=range(size), columns=self.topological_order)
+        sampled['_weight'] = [1] * size
+        for node in self.topological_order:
+            cpd = self.cpds[node]
+            if cpd.evidence:
+                weights = []
+                for i in range(size):
+                    evid = [sampled[var][i] for var in cpd.evidence]
+                    weights.append(cpd.reduce(evid, inplace=False).values)
+                if node in evidence:
+                    sampled[node] = [evidence[node]] * size
+                    for i in range(size):
+                        sampled.loc[i, '_weight'] *= weights[i][evidence[node].state]
+                else:
+                    sampled[node] = sample_discrete(cpd.variables[cpd.variable], weights)
+            else:
+                weights = cpd.values
+                if node in evidence:
+                    sampled[node] = [evidence[node]] * size
+                    for i in range(size):
+                        sampled.loc[i, '_weight'] *= weights[evidence[node].state]
+                else:
+                    sampled[node] = sample_discrete(cpd.variables[cpd.variable], weights, size)
+        return sampled
+
+
+class MarkovChainMonteCarlo(nx.DiGraph):
```

Comment:
@abinashpanda One issue that arises when I inherit from `DirectedGraph` is: 
on calling `self.add_weighted_edges_from`, it uses the definition in `nx.Graph.add_weighted_edges_from`, which is simply `self.add_edges_from(((u,v,{weight:d}) for u,v,d in ebunch),**attr)`. But `add_edges_from` has been overridden by the definition `DirectedGraph.add_edges_from`, which in turn calls `DirectedGraph.add_edge` - but this does not support passing an argument for the weight, and thus results in an error. I could modify the definition of `DirectedGraph.add_edge` a bit, would you suggest that?